### PR TITLE
add API for order reprocess

### PR DESCRIPTION
**Pull request by AlexanderPavlenko**

### CODE REVIEW

File: lib/straight-server/order.rb [hunk 1]
```diff
@@ -73,31 +73,41 @@ def self.find_by_address(address)
...
 
       confirmations_required = gateway.confirmations_required
 
       transactions = gateway.fetch_transactions_for(address)
-      transactions = Straight::Transaction.from_hashes transactions
+      transactions = Straight::Transaction.from_hashes(transactions)
       transactions.select! { |x| x.confirmations >= confirmations_required } if confirmations_required > 0
 
-      same_address_orders = Order.exclude(id: id).where(gateway_id: gateway.id, address: self.address)
+      same_address_orders = self.class.exclude(id: id).where(gateway_id: gateway_id, address: address)
```
Comment (AlexeyShepelev) [inline lib/straight-server/order.rb]: Why do you dislike a class name?

File: lib/straight-server/order.rb [hunk 2]
```diff
@@ -73,31 +73,41 @@ def self.find_by_address(address)
...
       end
 
       result = get_transaction_status(transactions: transactions)
-      return if result[:status] == 0
+      return false if result[:status] == 0
       result[:status] = 3 if result[:status] == -3
 
-      if result[:status] != @status
+      if (result[:status] != @status) || (result[:amount_paid] != amount_paid)
+        @old_status = @status
```
Comment (AlexeyShepelev) [inline lib/straight-server/order.rb]: Where do you use `@old_status`?

File: lib/straight-server/order.rb (hunk 1 above)
Comment (AlexanderPavlenko) [inline lib/straight-server/order.rb]: Not so important here, but explicit reference may conflict with inheritance, and there's an option for it, `gateway.order_class` field: https://github.com/MyceliumGear/straight-server/blob/master/lib/straight-server/gateway.rb#L611

File: lib/straight-server/order.rb [hunk 3]
```diff
@@ -73,31 +73,41 @@ def self.find_by_address(address)
...
+        sao_block_heights = same_address_orders.select_map(:block_height_created_at)
+        if sao_block_heights.index { |x| x.to_i <= 0 }
+          raise "Ambiguity detected: same-address order with undefined block_height_created_at"
+        end
+        if sao_block_heights.index { |x| x == block_height_created_at }
+          raise "Ambiguity detected: same-address order with the same block_height_created_at"
+        end
+        max_allowed_block_height = sao_block_heights.min
+        transactions.select! { |x| (x.block_height.to_i > 0) && (x.block_height <= max_allowed_block_height) }
       end
```
Comment (AlexanderPavlenko) [inline lib/straight-server/order.rb]: Yep. Also, looks like these are not fully covered with spec and methods extraction may help.

File: lib/straight-server/order.rb (hunk 2 above)
Comment (AlexanderPavlenko) [inline lib/straight-server/order.rb]: https://github.com/MyceliumGear/straight-server/blob/1343fa12c9ad693337237d996c5346eb28ae15c4/lib/straight-server/gateway.rb#L212

File: lib/straight-server/order.rb [hunk 4]
```diff
@@ -73,31 +73,41 @@ def self.find_by_address(address)
...
+        @old_status = @status
         @status = self[:status] = result[:status]
         self.amount_paid = result[:amount_paid]
         self.accepted_transactions = result[:accepted_transactions]
         save
 
-        gateway.order_status_changed(self)
+        Thread.new do
+          gateway.order_status_changed(self)
+        end
```
Comment (AlexanderPavlenko) [inline lib/straight-server/order.rb]: They go to the threaven :)
http://stackoverflow.com/a/4100026

Comment (AlexanderPavlenko) [inline lib/straight-server/order.rb]: Alive threads are more a problem, there can become too many of them.

**Pull request merged**